### PR TITLE
[WIP][CONNECT] Test `RootAllocator` memory leak

### PR DESCRIPTION
**Pull request by LuciferYang**

### CODE REVIEW

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala [hunk 1]
```diff
@@ -117,6 +117,7 @@ trait RemoteSparkSession
   override def beforeAll(): Unit = {
     super.beforeAll()
     SparkConnectServerUtils.start()
+    System.setProperty("arrow.memory.debug.allocator", false.toString)
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:120]: If `arrow.memory.debug.allocator` is true, 

https://github.com/apache/arrow/blob/798b417ec342051f3ff30b9953dcb1f1b71dc9e8/java/memory/memory-core/src/main/java/org/apache/arrow/memory/BaseAllocator.java#L428-L438

The memory leak may be covered, let's set "arrow.memory.debug.allocator" to false and use GA check it

Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:120]: run `build/sbt "connect-client-jvm/testOnly *ClientE2ETestSuite"` with this change:

```
[info] ClientE2ETestSuite:
[info] - test spark result schema (58 milliseconds)
[info] - test spark result array (42 milliseconds)
[info] - simple dataset test (210 milliseconds)
[info] - simple udf test (252 milliseconds)
[info] org.apache.spark.sql.ClientE2ETestSuite *** ABORTED *** (8 seconds, 283 milliseconds)
[info]   "Memory was leaked by query. Memory leaked: (272)
[info]   Allocator(ROOT) 0/272/272/9223372036854775807 (res/actual/peak/limit)
[info]   " contained "Memory was leaked by query. Memory leaked" (RemoteSparkSession.scala:160)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
[info]   at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
[info]   at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
[info]   at org.apache.spark.sql.connect.client.util.RemoteSparkSession.afterAll(RemoteSparkSession.scala:160)
[info]   at org.apache.spark.sql.connect.client.util.RemoteSparkSession.afterAll$(RemoteSparkSession.scala:155)
[info]   at org.apache.spark.sql.ClientE2ETestSuite.afterAll(ClientE2ETestSuite.scala:23)
[info]   at org.scalatest.BeforeAndAfterAll.$anonfun$run$1(BeforeAndAfterAll.scala:225)
[info]   at org.scalatest.Status.$anonfun$withAfterEffect$1(Status.scala:377)
[info]   at org.scalatest.Status.$anonfun$withAfterEffect$1$adapted(Status.scala:373)
[info]   at org.scalatest.CompositeStatus.whenCompleted(Status.scala:962)
[info]   at org.scalatest.Status.withAfterEffect(Status.scala:373)
[info]   at org.scalatest.Status.withAfterEffect$(Status.scala:371)
[info]   at org.scalatest.CompositeStatus.withAfterEffect(Status.scala:863)
[info]   at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:224)
[info]   at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]   at org.apache.spark.sql.ClientE2ETestSuite.run(ClientE2ETestSuite.scala:23)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:321)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:517)
[info]   at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:413)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]   at java.lang.Thread.run(Thread.java:750)
[info] Run completed in 9 seconds, 695 milliseconds.
[info] Total number of tests run: 4
[info] Suites: completed 0, aborted 1
[info] Tests: succeeded 4, failed 0, canceled 0, ignored 0, pending 0
[info] *** 1 SUITE ABORTED ***
[error] Error during tests:
[error] 	org.apache.spark.sql.ClientE2ETestSuite

```

**Comment by hvanhovell** [comment]
cc @zhenlineo we discussed this today

**Comment by zhenlineo** [comment]
Thanks for looking into this. We should definitely fix the leak. I would probably do the same:
When we close the session, we go through all buffers and close them. The Result and cleaner also holds some resources.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 2]
```diff
@@ -133,6 +133,7 @@ class SparkSession(private val client: SparkConnectClient, private val cleaner:
 
   override def close(): Unit = {
     client.shutdown()
+    cleaner.stop()
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:136]: Like this? stop cleaner and clean up close ref in `referenceBuffer` that has not been cleaned? @zhenlineo

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala (hunk 1 above)
Comment (amaliujia) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:122]: What is GA test BTW?

Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/util/RemoteSparkSession.scala:122]: GitHub Action ...

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 2 above)
Comment (zhenlineo) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala:136]: I am looking at this afternoon. We cannot directly shutdown the cleaner as it is shared by all sessions. I have some draft [here](https://github.com/zhenlineo/spark/pull/new/allocator-memleak). Be free to use any idea there. If you do not have more time on this. I will add some testing and send a PR tomorrow.

**Comment by hvanhovell** [comment]
We can also just clean-up the results. Part of the problem is that the E2E test does not close results. I am not sure if there are similar issues on the server side.

**Comment by hvanhovell** [comment]
Will submit a PR shortly.

**Comment by zhenlineo** [comment]
@hvanhovell What about this? https://github.com/apache/spark/compare/master...zhenlineo:spark:allocator-memleak?expand=1 when a session is closed, we force to close all buffers.

**Comment by LuciferYang** [comment]
Let me close this one, thanks @hvanhovell @zhenlineo

**Pull request closed without merging**